tfx/orchestration/experimental/optimizations/beam_fusion.py

```diff
@@ -0,0 +1,197 @@
+       parents' ancestors) should be in the current subgraph being explored.
+
+    Additionally, each subrgaph must meet the following optimality constraint:
+    1. There can be no subgraphs such that their union satisfies 1, 2, and 3
+       as described above.
+    """
+
+    # Finds subgraphs of fuseable beam components through a BFS of each source.
+    candidate_sources = self.get_subgraph_sources(self.pipeline.components)
+    fuseable_subgraphs = []
+    visited = set()
+
+
+    for source in sorted(candidate_sources, key=lambda c: c.id): # pylint: disable=too-many-nested-blocks
+      if source in visited:
+        continue
+
+      subgraph = []
+      current_layer = [source]
+
+      # Conduct a BFS on the candidate source.
+      while current_layer:
+        next_layer = []
+        for component in current_layer:
+          if component in visited:
+            continue
+
+          visited.add(component)
+          subgraph.append(component)
+
+          # Iterate through the children to find Beam components
+          for child in component.downstream_nodes:
+            if not self._is_beam_component(child):
+              continue
+
+            # Checks if the child is in an explored subgraph that needs to be
+            # fused into the current subrgaph
+            if child in visited:
+              intersecting_subgraph = self._get_intersecting_subgraph(
+                  child, fuseable_subgraphs)
+              if intersecting_subgraph:
+                subgraph.extend(intersecting_subgraph)
+                fuseable_subgraphs.remove(intersecting_subgraph)
+
+            elif self._is_fuseable(child, subgraph):
+              next_layer.append(child)
+
+          current_layer = next_layer
+
+      if len(subgraph) > 1:
+        fuseable_subgraphs.append(subgraph)
+
+    # Topologically sort all the subgraphs
+    for i in range(len(fuseable_subgraphs)): # pylint: disable=consider-using-enumerate
+      fuseable_subgraphs[i] = self._topologically_sort(fuseable_subgraphs[i],
+                                                       candidate_sources)
+    return fuseable_subgraphs
+
+  def get_subgraph_sources(self, subgraph: List[base_node.BaseNode]):
+    """Finds sources (components with no parents that are Apache Beam based components)"""
+    subgraph_sources = set()
+
+    for component in subgraph:
+      if not self._is_beam_component(component):
+        continue
+
+      is_source = True
+
+      for parent in component.upstream_nodes:
+        if self._is_beam_component(parent):
+          is_source = False
+          break
+
+      if is_source:
+        subgraph_sources.add(component)
+
+    return subgraph_sources
+
+  def get_subgraph_sinks(self, subgraph: List[base_node.BaseNode]):
+    """Finds sinks (components with no children that are in the subgraph)."""
+    subgraph_sinks = set()
+
+    for component in subgraph:
+      is_sink = True
```

Comment:
I initially assumed it would only be called on subgraphs with all beam components, but I added the `_is_beam_component` to `get_subgraph_sinks` as an extra safety check.